### PR TITLE
Use setup.py for requirements.txt dependencies.

### PR DESCRIPTION
**Pull request by schmmd**
I noticed our `requirements.txt` and `setup.py` diverging, and then found this neat trick from https://caremad.io/posts/2013/07/setup-vs-requirement/.

The TeamCity build will verify whether this works correctly.

### CODE REVIEW

**Comment by schmmd** [comment]
+1 for TeamCity, as it correctly identified this as a breaking change where Travis did not.

**Comment by schmmd** [comment]
On the Docker image tests pass, but pylint fails pretty terribly.

**Comment by schmmd** [comment]
It's using different settings.  I.e. in the logs of this PR I get`C: 18, 0: Line too long (105/100) (line-too-long)` but locally the line length limit is 115.

**Comment by schmmd** [comment]
I do not understand how pylint has been working successfully on TeamCity (as it seems to have been) to date, as the file I added in the previous commit was missing.

**Comment by schmmd** [comment]
I fixed the issue I had been having with https://github.com/allenai/allennlp/pull/638.

**Comment by schmmd** [comment]
There is one functional change in that `pip install -r requirements.txt` will install `allennlp` into `pip`.  It will be importable from directories other than the base `allennlp` directory.

On the upside, we don't need to duplicate `requirements.txt` and `setup.py`, which has been causing issues with our pip release.

File: setup.py [hunk 1]
```diff
@@ -85,26 +85,47 @@
...
+          'numpy',
+          # aws commandline tools for running on Docker remotely.
           'awscli>=1.11.91',
+          # REST interface for models
           'sanic==0.6.0',
+          'sanic-cors==0.6.0.2',
+          # Talk to postgres demo database
           'psycopg2',
-          'sanic-cors',
+          # argument parsing in for data cleaning scripts
```
Comment (matt-gardner) [inline setup.py]: Typo was present in `requirements.txt`, but can you fix it?  `s/in for/for/`.

**Comment by matt-gardner** [comment]
Oh, wait a minute, I'm not certain that's a good thing...  I'd prefer to not install the library when running `pip install -r requirements.txt` - I'd rather there not be any ambiguity in my development environment when I'm developing allennlp.

**Comment by schmmd** [comment]
@matt-gardner I agree--I would also prefer not to install `allennlp` as a library.  However, I think this has a net positive benefit of avoiding duplicating our dependency versions and others take this approach as well.

**Comment by schmmd** [comment]
I don't think the rules are ambiguous: if you are in the `allennlp` directory you will be using the source code.

**Comment by joelgrus** [comment]
I really don't like having allennlp installed as a library in the virtualenv you're working on source in

**Comment by matt-gardner** [comment]
Neither does @DeNeutoy: https://github.com/allenai/allennlp/pull/627#pullrequestreview-83945500.  Is there no other way to remove this duplication?

**Comment by DeNeutoy** [comment]
The issue with this is that there _is_ a subtle distinction between `requirements.txt` and `setup.py` (or at least in how I have interpreted their use), specifically that `requirements.txt` should be for development and `setup.py` for use as a library. The other thing enforcing this separation is that you can't have dependencies pinned to git commits in setup.py (it breaks the install), but this way around, we can't temporarily pin libraries to git commits because they are installed from `setup.py`.
[This probably describes it better.](https://packaging.python.org/discussions/install-requires-vs-requirements/#install-requires-vs-requirements-files)

**Comment by DeNeutoy** [comment]
Additionally, this will be unnecessary once we have a TeamCity job which builds a wheel and runs a subset of our tests, as this would probably catch most critical missing dependencies.

**Comment by joelgrus** [comment]
I can write a python script that checks that the two files are in sync and we can add it to our CI, the same way that we already do for our docs.

**Comment by schmmd** [comment]
@DeNeutoy we somewhat remove the "subtle distinction" by separating out our development dependencies into `requirements_test.txt`.  I think avoiding dependency duplication would still be good even when we have more tests.

@joelgrus good idea on writing a CI job (although I still think the duplication between `setup.py` and `requirements.txt` is a bit nuts.

**Pull request closed without merging**